### DISCUSSION TITLE
Trying to compile 3.0.0 with CasADi

### DISCUSSION THREAD

**Discussion by mihirus** [stack-of-tasks/pinocchio#2280] [Q&A] [answered]
I'm trying to compile 3.0.0 from source with casadi. I followed the "Build from source" page from this guide: https://stack-of-tasks.github.io/pinocchio/download.html but I think it did not install the casadi bindings because I would get a `ModuleNotFoundError` when I tried to import `pinocchio.casadi`. They were also not present in the `dist-packages` folder, and I also verified that 3.0.0 was installed with `pinocchio.__version__`.

I found [2251](https://github.com/stack-of-tasks/pinocchio/issues/2251) which explains the various compile options and [2223](https://github.com/stack-of-tasks/pinocchio/discussions/2223) which gave the explicit cmake command to use these options.

I ran `cmake .. -DBUILD_WITH_CASADI_SUPPORT=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr/local` and got the following error. How to proceed from here?

![image](https://github.com/stack-of-tasks/pinocchio/assets/22569675/7f15ccb7-167c-43e6-bbb2-eeaaa2c4f12b)

**Comment by jorisv** [comment, marked as answer]
Hello,

You can look at [root CMakeLists.txt](https://github.com/stack-of-tasks/pinocchio/blob/master/CMakeLists.txt#L90-L139) to find all possible CMake options. You can also use cmake-gui to have a nice view of them.

How and where did you install Casadi ? Apparently CMake don't find the Casadi finder that help CMake to know where Casadi lib and include are installed.

You can get a look at the [Linux workflow](https://github.com/stack-of-tasks/pinocchio/blob/master/.github/workflows/linux.yml#L73-L131) that install casadi from robotpkg. This workflow is building pinocchio 3 with casadi on Ubuntu 20 and 22.

**Reply by mihirus** [reply]
Thank you! I had previously installed casadi via pip, but once I installed it via robotpkg I was able to build.

My next question is how do I get the python bindings to work? I see that the dynamic libraries have been built and installed to my  `/usr/local/lib`:
![image](https://github.com/stack-of-tasks/pinocchio/assets/22569675/6e67e8c9-21f8-4ed4-95ca-247c3280f145)

However, I cannot `import pinocchio` in python yet which I'm guessing is because my bindings are not configured properly. I tried using [these instructions](https://github.com/conda-forge/pinocchio-feedstock#installing-pinocchio) but it looks like it installs 2.7.1 bindings.

**Reply by nim65s** [reply]
What do you have in `/usr/local/lib/python3.10` ? any `site-package` or `dist-package` directory ? And what do you have inside that ?

Then, when you run `python -c "import sys; print(sys.path)"`, do you have any directory that looks like what you have in that `/usr/local/lib/pinocchio3.10` prefix ?

**Reply by mihirus** [reply]
`/usr/local/lib/python3.10/dist-packages` exists on my system but is empty. When I print `sys.path` the `/usr/local/lib/python3.10/dist-packages` does indeed show up in the list of paths which I've given below for reference:

`['', '/opt/drake/lib/python3.10/site-packages', '/opt/ros/humble/lib/python3.10/site-packages', '/opt/ros/humble/local/lib/python3.10/dist-packages', '/opt/openrobots/lib/python3.10/site-packages', '$HOME', '/usr/lib/python310.zip', '/usr/lib/python3.10', '/usr/lib/python3.10/lib-dynload', '$HOME/.local/lib/python3.10/site-packages', '/usr/local/lib/python3.10/dist-packages', '/usr/lib/python3/dist-packages', '/usr/lib/python3.10/dist-packages']`

**Reply by nim65s** [reply]
python bindings should be installed in the same prefix as dynamic libraries. if you can't find any file that looks like `pinocchio_pywrap.cpython-310-x86_64-linux-gnu.so` in that prefix, it probably means you compiled pinocchio without python bindings.

Can you double check that `BUILD_PYTHON_INTERFACE` is `ON`  in your configure folder ? And maybe try to reinstall pinocchio, and look at CMake logs to search for that `pinocchio_pywrap` file ?

**Reply by mihirus** [reply]
I found the pywrap and other files in `build/bindings/python/pinocchio`. I copied this entire `pinocchio` folder over to `/usr/local/lib/python3.10/dist-packages` and now I am able to import both `pinocchio` and `pinocchio.casadi`!

For the sake of understanding, was there anything I should have done after `sudo make install` to install the python bindings?

**Reply by nim65s** [reply]
`sudo make install` should be enough. You can run that again, and check for those files in the installation logs.
